docs/reference/c++_support.rst

```diff
@@ -0,0 +1,173 @@
+.. meta::
+  :description: This chapter describes the C++ support of the HIP ecosystem
+  ROCm software.
+  :keywords: AMD, ROCm, HIP, C++
+
+*******************************************************************************
+C++ language support
+*******************************************************************************
+
+The ROCm platform enables the power of combined C++ and HIP (Heterogeneous-computing
+Interface for Portability) code. This code is compiled with a ``clang`` or ``clang++``
+compiler. The official compilers support the HIP platform, but for the most up-to-date
+feature set, use the ``amdclang`` or ``amdclang++`` components included in the ROCm
+installation.
+
+The source code is compiled according to the ``C++03``, ``C++11``, ``C++14``, ``C++17``,
+and ``C++20`` standards, along with HIP-specific extensions, but is subject to
+restrictions. The key restriction is the lack of standard library support. This is due to
+the HIP device's Single Instruction/Multiple Data (SIMD) nature, which makes most of the
+standard library implementations not performant or useful. However, essential operations
+are implemented in HIP-specific libraries such as `rocPRIM
+<https://github.com/ROCm/rocprim>`_, `rocThrust <https://github.com/ROCm/rocthrust/>`_,
+and `hipCUB <https://github.com/ROCm/hipcub/>`_.
+
+.. _language_modern_c++_support:
+Modern C++ support
+===============================================================================
+
+C++ is considered a modern programming language as of C++11. This section describes how
+HIP supports these new C++ features.
+
+C++11 support
+-------------------------------------------------------------------------------
+
+The C++11 standard introduced many new features. These features are supported in HIP
+device code, with some notable omissions. The most significant is the lack of concurrency
+support on the device. This is because the HIP device concurrency model fundamentally
```

Comment:
It is important to note that although C++11 language features are supported by HIP for both host and device code, C++11 standard library features are mostly supported for host code only. The implementation of the C++11 standard library does not consider being used by HIP device code. Even though constexpr functions are implicitly host-device functions that can be called by HIP device code, the implementation of certain C++11 standard library features involves using non-constexpr functions or static members, making them unavailable for HIP device code (e.g., std::function). Therefore, C++11 standard library features are not officially supported by HIP for device code and are in a state of "work as-is."